crypto/encode_decode/encoder_lib.c

```diff
@@ -312,120 +369,293 @@ OSSL_ENCODER_INSTANCE_get_output_type(OSSL_ENCODER_INSTANCE *encoder_inst)
     return encoder_inst->output_type;
 }
 
-static int encoder_process(OSSL_ENCODER_CTX *ctx, BIO *out)
+const char *
+OSSL_ENCODER_INSTANCE_get_output_structure(OSSL_ENCODER_INSTANCE *encoder_inst)
 {
-    size_t i, end;
-    void *latest_output = NULL;
-    size_t latest_output_length = 0;
-    const char *latest_output_type = NULL;
-    const char *last_input_type = NULL;
-    int ok = 0;
-
-    end = OSSL_ENCODER_CTX_get_num_encoders(ctx);
-    for (i = 0; i < end; i++) {
-        OSSL_ENCODER_INSTANCE *encoder_inst =
-            sk_OSSL_ENCODER_INSTANCE_value(ctx->encoder_insts, i);
-        OSSL_ENCODER *encoder = OSSL_ENCODER_INSTANCE_get_encoder(encoder_inst);
-        void *encoderctx = OSSL_ENCODER_INSTANCE_get_encoder_ctx(encoder_inst);
-        const char *current_input_type =
-            OSSL_ENCODER_INSTANCE_get_input_type(encoder_inst);
-        const char *current_output_type =
-            OSSL_ENCODER_INSTANCE_get_output_type(encoder_inst);
-        BIO *current_out;
-        BIO *allocated_out = NULL;
-        const void *current_data = NULL;
-        OSSL_PARAM abstract[3];
-        OSSL_PARAM *abstract_p;
-        const OSSL_PARAM *current_abstract = NULL;
-
-        if (latest_output_type == NULL) {
-            /*
-             * This is the first iteration, so we prepare the object to be
-             * encoded
-             */
+    if (encoder_inst == NULL)
+        return NULL;
+    return encoder_inst->output_structure;
+}
 
-            current_data = ctx->construct(encoder_inst, ctx->construct_data);
+static int encoder_process(struct encoder_process_data_st *data)
+{
+    OSSL_ENCODER_INSTANCE *current_encoder_inst = NULL;
+    OSSL_ENCODER *current_encoder = NULL;
+    OSSL_ENCODER_CTX *current_encoder_ctx = NULL;
+    BIO *allocated_out = NULL;
+    const void *original_data = NULL;
+    OSSL_PARAM abstract[10];
+    const OSSL_PARAM *current_abstract = NULL;
+    int i;
+    int ok = -1;  /* -1 signifies that the lookup loop gave nothing */
+    int top = 0;
+
+    if (data->next_encoder_inst == NULL) {
+        /* First iteration, where we prepare for what is to come */
+
+        data->count_output_structure =
+            data->ctx->output_structure == NULL ? -1 : 0;
+        top = 1;
+    }
 
-            /* Assume that the constructor recorded an error */
-            if (current_data == NULL)
-                goto loop_end;
+    for (i = data->current_encoder_inst_index; i-- > 0;) {
```

Comment:
So this looks like the order of the encoder list is fairly important..
It goes backwards thru the list on the first level until it gets a match.. then it recursively calls, but the index it uses starts from the match point and goes backwards from there..
So a chain of multiple levels would have to obey the order rule also..

Reply:
Yes.  Was this just a remark, like thinking out loud to get confirmation, or are you trying to say something else between the lines?

Reply:
The observation is that is a fairly nasty depandancy. Someone adding to the encoder list would probably not know this right?

Reply:
Yeah...  so if you want to implement your own OSSL_ENCODER_CTX_new_by_{customtype}(), you have to know what you're doing.  I have a separate (and currently unpublished) branch that refactors the OSSL_ENCODER and OSSL_DECODER documentation to emphasize and clarify this.  I plan on submitting that after this and #13248 have been merged.

Reply:
The expectation is, though, that `OSSL_ENCODER_CTX_add_encoder() is used to add encoders from the type you want to encode to whatever the next step is (DER output, for example), and that anything that needs added to get the chains (such as DER to PEM) is added with `OSSL_ENCODER_CTX_add_extra()`, which is automated.  As long as the stack of encoders is set up that way, it's pretty safe.

Reply:
I am probably just not understanding something..
The currently implementation is complicated by the chaining support. The main usage for the chaining seems to be for the DER to PEM usage which doesnt seem to be part of this PR???
Is all this extra complexity just to handle one case really needed? Yes we can talk about future extensions - but why overcomplicate it for something that might not be needed.

Reply:
native->der->pem is a known use-case.  what other uses are there -- real, or possible?

Can the "BIO pipeline" concept be used here as well, to simplify the design?